### ISSUE TITLE
Possible race in systemd-homed/TEST-46-HOMED

### ISSUE THREAD

**Issue by mrc0mmand** [systemd/systemd#21589]
**systemd version the issue has been seen with**
 > latest main

**Used distribution**
 > Arch Linux

In last ~three days I saw a couple (literally 2) instances of TEST-46-HOMED failing in a way I haven't seen so far:

```
[   87.854973] testsuite-46.sh[267]: + homectl with test-user -- touch /home/test-user/xyz
[   87.855741] systemd-homed[240]: test-user2: changing state active → resizing-while-active
[   87.878612] systemd-homed[240]: test-user: changing state inactive → activating-for-acquire
[   87.878973] systemd-homed[240]: Rebalancing complete.
[   87.940463] systemd-homework[1029]: Provided password unlocks user record.
[   87.941024] systemd-homework[1029]: Successfully locked image file '/home/test-user.home'.
[   88.088813] kernel: loop0: detected capacity change from 0 to 557056
[   87.948578] systemd-homework[1029]: Setting up loopback device /dev/loop0 completed.
...
[   90.105502] kernel: BTRFS info (device dm-0): resize device /dev/mapper/home-test-user (devid 1) from 292884480 to 268435456
[   90.115529] kernel: dm-0: detected capacity change from 572040 to 524288
[   90.130615] kernel: BTRFS info (device dm-0): devid 1 device path /dev/mapper/home-test-user changed to /dev/dm-0 scanned by 6 (519)
[   90.142123] kernel: BTRFS info (device dm-0): devid 1 device path /dev/dm-0 changed to /dev/mapper/home-test-user scanned by 6 (519)
[   90.020820] systemd-homework[1068]: LUKS device shrinking completed.
[   90.080356] systemd-homework[1068]: Refreshing loop device size completed.
[   90.080583] systemd-homework[1068]: Shrinking of image file completed.
[   90.082241] systemd-homework[1068]: Shrinking of partition completed.
[   90.223732] kernel: loop0: detected capacity change from 604808 to 557056
[   90.086887] systemd-homework[1068]: Discarded unused 24.3M.
[   90.087123] systemd-homework[1068]: Synchronized disk.
[   90.087237] systemd-homework[1068]: Resizing completed.
[   90.087383] systemd-homework[1068]: Image size is 274.0M, file system size is 256.0M, file system payload size is 256.0M, file system free is 252.4M.
[   90.108027] systemd-homework[1068]: Unmounting completed.
[   90.108185] systemd-homework[1068]: Discovered used LUKS device /dev/mapper/home-test-user.
[   90.108276] systemd-homework[1068]: Device home-test-user is not active.
[   90.151432] systemd-homework[1068]: Everything completed.
[   90.151627] systemd-homed[240]: block device /sys/devices/virtual/block/dm-0 has been removed.
[   90.186528] testsuite-46.sh[267]: + PASSWORD=xEhErW0ndafV4s
[   90.186528] testsuite-46.sh[267]: + homectl with test-user -- test -f /home/test-user/xyz
[   90.187180] systemd-homed[240]: test-user: changing state deactivating → inactive
[   90.191237] systemd-homed[240]: test-user2: changing state active → resizing-while-active
[   90.346620] kernel: device-mapper: ioctl: remove_all left 1 open device(s)
[   90.208754] systemd-homed[240]: test-user: changing state inactive → activating-for-acquire
[   90.209035] systemd-homed[240]: Rebalancing complete.
[   90.268690] systemd-homework[1078]: Provided password unlocks user record.
[   90.269518] systemd-homework[1078]: Image file '/home/test-user.home' already locked, can't use.
[   90.270375] systemd-homed[240]: Activation failed: Address already in use
[   90.270831] systemd-homed[240]: test-user: changing state activating-for-acquire → inactive
[   90.270972] systemd-homed[240]: Got notification that all sessions of user test-user ended, deactivating automatically.
[   90.271214] systemd-homed[240]: Home test-user already deactivated, no automatic deactivation needed.
[   90.271603] homectl[1076]: Operation on home test-user failed: Home test-user is currently being used elsewhere.
[   90.511241] systemd-homework[1077]: Discovered used LUKS device /dev/mapper/home-test-user2, and validated password.
[   90.640727] systemd-homework[1077]: Successfully re-activated LUKS device.
[   90.640961] systemd-homework[1077]: Discovered used loopback device /dev/loop1.
[   90.641078] systemd-homework[1077]: offset = 1048576, size = 818565120, image = 820662272
[   90.641221] systemd-homework[1077]: Grow operation would shrink file system, skipping operation.
[   90.644230] systemd-homed[240]: test-user2: changing state resizing-while-active → active
```

Full journals:
 * [system.journal.tar.gz](https://github.com/systemd/systemd/files/7632824/system.journal.tar.gz)
 * [system.journal.1.tar.gz](https://github.com/systemd/systemd/files/7632837/system.journal.1.tar.gz)

/cc @poettering

**Comment by keszybz** [comment]
@poettering could you take a look?

**Comment by keszybz** [comment]
#21253 was merged on Nov 25, I'd guess that it is the most likely culprit.
I'm not familiar with that code and I didn't even review #21253… Maybe somebody else want to take a look.

**Comment by yuwata** [comment]
Hm, the workaround de27f4c8b8dd230a936f1b467a1da275f6fe024b seems not work. See https://jenkins-systemd.apps.ocp.ci.centos.org/job/upstream-vagrant-archlinux/9174/artifact//systemd-centos-ci/index.html.

**Comment by mrc0mmand** [comment]
Yup, the issue is still present:

```
[  111.137593] systemd-homework[1318]: Discovered used LUKS device /dev/mapper/home-test-user2, and validated password.
[  111.265588] systemd-homework[1318]: Successfully re-activated LUKS device.
[  111.265820] systemd-homework[1318]: Discovered used loopback device /dev/loop1.
[  111.265918] systemd-homework[1318]: offset = 1048576, size = 818565120, image = 820662272
[  111.266049] systemd-homework[1318]: Grow operation would shrink file system, skipping operation.
[  111.271715] systemd-homed[245]: Change operation of test-user2 completed.
[  111.271973] systemd-homed[245]: test-user2: changing state resizing-while-active → active
[  111.272081] systemd-homed[245]: Scheduled immediate rebalancing...
[  111.272177] systemd-homed[245]: Scheduled immediate rebalancing...
[  111.272268] systemd-homed[245]: Unable to read dirty xattr off image file, ignoring: Operation not supported
[  111.272381] systemd-homed[245]: Rebalancing now...
[  111.272486] systemd-homed[245]: Rebalancing complete.
[  111.272582] systemd-homed[245]: Scheduled rebalancing in 4min 23.499694s...
[  112.029024] testsuite-46.sh[270]: + break
[  112.029024] testsuite-46.sh[270]: + homectl remove test-user
[  112.040543] systemd-homed[245]: Unable to read dirty xattr off image file, ignoring: Operation not supported
[  112.040637] systemd-homed[245]: Sending to worker: {"luksDiscard":true,"luksPbkdfType":"pbkdf2","luksPbkdfTimeCostUSec":1000,"userName":"test-user","disposition":"regular","lastPasswordChangeUSec":16403949>
[  112.040991] systemd-homed[245]: Successfully forked off '(sd-homework)' as PID 1320.
[  112.041121] systemd-homed[245]: Unable to read dirty xattr off image file, ignoring: Operation not supported
[  112.041216] systemd-homed[245]: test-user: changing state inactive → removing
[  112.057015] systemd-homework[1320]: Everything completed.
[  112.058215] systemd-homed[245]: Removal of test-user completed.
[  112.060336] systemd-homed[245]: Scheduled immediate rebalancing...
[  112.060860] testsuite-46.sh[270]: + systemd-detect-virt -cq
[  112.061138] systemd-homed[245]: Rebalancing now...
[  112.061253] systemd-homed[245]: Shrinking phase..
[  112.061400] systemd-homed[245]: Disk space usage by all home directories to rebalance: 17.4M — available disk space: 1.4G
[  112.061525] systemd-homed[245]: Rebalancing interval set to 4min 23.499s.
[  112.061645] systemd-homed[245]: Growing phase..
[  112.061910] systemd-homed[245]: Disk space usage by all home directories to rebalance: 17.4M — available disk space: 1.4G
[  112.062044] systemd-homed[245]: Rebalancing home directory 'test-user2' 782.6M → 1.2G.
[  112.062159] systemd-homed[245]: Rebalancing interval set to 4min 23.499s.
[  112.062159] systemd-homed[245]: Rebalancing interval set to 4min 23.499s.
[  112.077855] testsuite-46.sh[270]: + wait_for_state test-user2 active
[  112.077855] testsuite-46.sh[270]: + (( i=0 ))
[  112.077855] testsuite-46.sh[270]: + (( i<10 ))
[  112.081755] testsuite-46.sh[1322]: + homectl inspect test-user2
[  112.082396] systemd-homed[245]: Sending to worker: {"luksDiscard":true,"luksPbkdfType":"pbkdf2","luksPbkdfTimeCostUSec":1000,"userName":"test-user2","disposition":"regular","lastPasswordChangeUSec":1640395>
[  112.082677] testsuite-46.sh[1323]: + grep -qF 'State: active'
[  112.083180] systemd-homed[245]: Successfully forked off '(sd-homework)' as PID 1324.
[  112.083318] systemd-homed[245]: test-user2: changing state active → resizing-while-active
...
[  112.367939] testsuite-46.sh[270]: + sleep .5
[  112.872617] testsuite-46.sh[270]: + (( i++ ))
[  112.873068] testsuite-46.sh[270]: + (( i<10 ))
[  112.874724] testsuite-46.sh[1327]: + grep -qF 'State: active'
[  112.875948] testsuite-46.sh[1326]: + homectl inspect test-user2
...
[  114.101938] testsuite-46.sh[270]: + sleep .5
[  114.604963] testsuite-46.sh[270]: + (( i++ ))
[  114.604963] testsuite-46.sh[270]: + (( i<10 ))
[  114.605358] testsuite-46.sh[1329]: + homectl inspect test-user2
[  114.605875] testsuite-46.sh[1330]: + grep -qF 'State: active'
...
[  115.078222] testsuite-46.sh[270]: + sleep .5
[  115.581248] testsuite-46.sh[270]: + (( i++ ))
[  115.581248] testsuite-46.sh[270]: + (( i<10 ))
[  115.581924] testsuite-46.sh[1332]: + homectl inspect test-user2
[  115.582254] testsuite-46.sh[1333]: + grep -qF 'State: active'
...
[  116.265328] testsuite-46.sh[270]: + sleep .5
[  116.768447] testsuite-46.sh[270]: + (( i++ ))
[  116.768447] testsuite-46.sh[270]: + (( i<10 ))
[  116.769178] testsuite-46.sh[1335]: + homectl inspect test-user2
[  116.769692] testsuite-46.sh[1336]: + grep -qF 'State: active'
...
[  117.685078] testsuite-46.sh[270]: + sleep .5
[  118.187912] testsuite-46.sh[270]: + (( i++ ))
[  118.187912] testsuite-46.sh[270]: + (( i<10 ))
[  118.188702] testsuite-46.sh[1338]: + homectl inspect test-user2
[  118.189142] testsuite-46.sh[1339]: + grep -qF 'State: active'
...
[  119.352635] testsuite-46.sh[270]: + sleep .5
[  119.855985] testsuite-46.sh[270]: + (( i++ ))
[  119.855985] testsuite-46.sh[270]: + (( i<10 ))
[  119.856391] testsuite-46.sh[1341]: + homectl inspect test-user2
[  119.856994] testsuite-46.sh[1342]: + grep -qF 'State: active'
...
[  120.306797] testsuite-46.sh[270]: + sleep .5
[  120.810046] testsuite-46.sh[270]: + (( i++ ))
[  120.810046] testsuite-46.sh[270]: + (( i<10 ))
[  120.810748] testsuite-46.sh[1344]: + homectl inspect test-user2
[  120.811075] testsuite-46.sh[1345]: + grep -qF 'State: active'
...
[  121.498935] testsuite-46.sh[270]: + sleep .5
[  122.001513] testsuite-46.sh[270]: + (( i++ ))
[  122.002222] testsuite-46.sh[270]: + (( i<10 ))
[  122.003359] testsuite-46.sh[1347]: + homectl inspect test-user2
[  122.003870] testsuite-46.sh[1348]: + grep -qF 'State: active'
...
[  122.938262] testsuite-46.sh[270]: + sleep .5
[  123.441089] testsuite-46.sh[270]: + (( i++ ))
[  123.441958] testsuite-46.sh[270]: + (( i<10 ))
[  123.442987] testsuite-46.sh[1350]: + homectl inspect test-user2
[  123.443347] testsuite-46.sh[1351]: + grep -qF 'State: active'
...
[  124.161473] testsuite-46.sh[270]: + sleep .5
[  124.663834] testsuite-46.sh[270]: + (( i++ ))
[  124.664541] testsuite-46.sh[270]: + (( i<10 ))
[  124.664913] testsuite-46.sh[270]: + homectl deactivate test-user2
...
[  124.675558] homectl[1353]: Failed to deactivate user home: An operation on home test-user2 is currently being executed.
...
[  126.435213] sh[1355]: + systemctl poweroff --no-block
[  126.872375] systemd-logind[252]: System is powering down.
[  131.895526] systemd-homed[245]: Bus n/a: changing state RUNNING → CLOSED
[  131.896121] systemd-homed[245]: Worker process for home test-user2 is still running while exiting. Waiting for it to finish.
[  132.061687] systemd-homework[1324]: Discovered used LUKS device /dev/mapper/home-test-user2, and validated password.
[  132.199294] systemd-homework[1324]: Successfully re-activated LUKS device.
[  132.199615] systemd-homework[1324]: Discovered used loopback device /dev/loop1.
[  132.199760] systemd-homework[1324]: offset = 1048576, size = 818565120, image = 820662272
[  132.199890] systemd-homework[1324]: Ready to resize image size 782.6M → 800.0M, partition size 780.6M → 798.0M, file system size 764.6M → 782.0M.
[  132.200952] systemd-homework[1324]: Growing of image file from 782.6M to 800.0M completed.
[  132.235401] systemd-homework[1324]: Refreshing loop device size completed.
[  132.236604] systemd-homework[1324]: Growing of partition completed.
[  132.361072] kernel: loop1: detected capacity change from 1598760 to 1634304
[  132.365719] kernel: audit: type=1131 audit(1640395095.996:128): pid=1 uid=0 auid=4294967295 ses=4294967295 msg='unit=serial-getty@ttyS0 comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? termi>
[  132.373649] kernel: dm-1: detected capacity change from 1565992 to 1601536
[  132.412012] kernel: BTRFS info (device dm-1): devid 1 device path /dev/mapper/home-test-user2 changed to /dev/dm-1 scanned by 6 (1367)
[  132.424064] kernel: BTRFS info (device dm-1): devid 1 device path /dev/dm-1 changed to /dev/mapper/home-test-user2 scanned by 6 (1367)
[  132.318113] systemd-homework[1324]: LUKS device growing completed.
[  132.319944] systemd-homework[1324]: File system resizing from 764.6M to 782.0M completed.
[  132.446785] kernel: BTRFS info (device dm-1): resize device /dev/mapper/home-test-user2 (devid 1) from 801787904 to 819986432
[  132.320639] systemd-homework[1324]: Synchronized disk.
[  132.322153] systemd-homework[1324]: Discarded unused 41.9M.
[  132.322352] systemd-homework[1324]: Synchronized disk.
[  132.323698] systemd-homework[1324]: Resizing completed.
[  132.323851] systemd-homework[1324]: Image size is 800.0M, file system size is 782.0M, file system payload size is 782.0M, file system free is 778.4M.
```

Full journal: [systemd.journal.tar.gz](https://github.com/systemd/systemd/files/7777334/systemd.journal.tar.gz)


It's weird it takes so long for systemd-homework to kick in, and it also looks like it needs systemd-homed to *force* it to do something?

**Comment by mrc0mmand** [comment]
I can't seem to reproduce this anymore, so I dropped the test from the flake list, hopefully CI agrees as well.
